File: 0648-replace-words/0648-replace-words.cs
public class Solution {
    public string ReplaceWords(IList<string> dictionary, string sentence) {
        HashSet<string> rootSet = new HashSet<string>(dictionary);
        StringReader sr = new StringReader(sentence);
        StringWriter sw = new StringWriter();
        string word;
        bool firstWord = true;

        while ((word = ReadWord(sr)) != null)
        {
            string prefix = null;
            for (int i = 1; i <= word.Length; ++i)
            {
                prefix = word.Substring(0, i);
                if (rootSet.Contains(prefix))
                {
                    word = prefix;
                    break;
                }
            }
            if (!firstWord)
            {
                sw.Write(" ");
            }
            sw.Write(word);
            firstWord = false;
        }

        return sw.ToString();
    }

    private string ReadWord(StringReader sr)
    {
        StringBuilder sb = new StringBuilder();
        int ch;
        while ((ch = sr.Read()) != -1)
        {
            if (char.IsWhiteSpace((char)ch))
            {
                if (sb.Length > 0)
                {
                    break;
                }
                else
                {
                    continue;
                }
            }
            sb.Append((char)ch);
        }
        return sb.Length > 0 ? sb.ToString() : null;
    }

}

File: 0648-replace-words/0648-replace-words.cpp
class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        unordered_set<string> rootSet(dictionary.begin(), dictionary.end());
        istringstream iss(sentence);
        ostringstream oss;
        string word;
        bool firstWord = true;

        while (iss >> word) {
            string prefix;
            for (size_t i = 1; i <= word.length(); ++i) {
                prefix = word.substr(0, i);
                if (rootSet.find(prefix) != rootSet.end()) {
                    word = prefix;
                    break;
                }
            }
            if (!firstWord) {
                oss << " ";
            }
            oss << word;
            firstWord = false;
        }

        return oss.str();
    }
};